# 03 — Hypothesis Testing (with effect sizes)
**RetainIQ India (Day 6).** At n=7,043 almost everything is 'significant', so every test is paired with an **effect size** — that is what tells us which drivers actually matter.

In [1]:
import sys; sys.path.insert(0,'../src')
import hypothesis_tests, pandas as pd
out = hypothesis_tests.run()
pd.DataFrame(out['tests'])

06:01:24 | INFO    | htest | [chi2] contract_type    chi2=  1184.6 p=5.86e-258 V=0.410 (large)


06:01:24 | INFO    | htest | [chi2] payment_method   chi2=   648.1 p=3.68e-140 V=0.303 (large)


06:01:24 | INFO    | htest | [chi2] internet_type    chi2=   732.3 p=9.57e-160 V=0.322 (large)


06:01:24 | INFO    | htest | [chi2] senior_citizen   chi2=   159.4 p=1.51e-36 V=0.151 (moderate)


06:01:24 | INFO    | htest | [t]    tenure_months    t=   -34.8 p=1.20e-232 d=-0.852 (large)  churn_mean=18.0 retain_mean=37.6


06:01:24 | INFO    | htest | [t]    monthly_charges_inr t=    18.4 p=8.59e-73 d=+0.446 (small)  churn_mean=74.4 retain_mean=61.3


06:01:24 | INFO    | htest | Drivers ranked by EFFECT SIZE:


06:01:24 | INFO    | htest |    tenure_months      Cohen's d    size=-0.852


06:01:24 | INFO    | htest |    monthly_charges_inr Cohen's d    size=0.446


06:01:24 | INFO    | htest |    contract_type      Cramer's V   size=0.410


06:01:24 | INFO    | htest |    internet_type      Cramer's V   size=0.322


06:01:24 | INFO    | htest |    payment_method     Cramer's V   size=0.303


06:01:24 | INFO    | htest |    senior_citizen     Cramer's V   size=0.151


,test,factor,stat,p,effect,effect_size,verdict
0,chi-square,contract_type,1184.6,5.863038e-258,Cramer's V,0.410,reject H0; effect large
1,chi-square,payment_method,648.1,3.682355e-140,Cramer's V,0.303,reject H0; effect large
2,chi-square,internet_type,732.3,9.571788e-160,Cramer's V,0.322,reject H0; effect large
3,chi-square,senior_citizen,159.4,1.510067e-36,Cramer's V,0.151,reject H0; effect moderate
4,Welch t-test,tenure_months,-34.8,1.195495e-232,Cohen's d,-0.852,reject H0; effect large
5,Welch t-test,monthly_charges_inr,18.4,8.592449e-73,Cohen's d,0.446,reject H0; effect small


## Drivers ranked by |effect size|

In [2]:
pd.DataFrame(out['tests']).assign(abs_effect=lambda d: d.effect_size.abs())\
  .sort_values('abs_effect', ascending=False)[['factor','test','effect','effect_size']]

,factor,test,effect,effect_size
4,tenure_months,Welch t-test,Cohen's d,-0.852
5,monthly_charges_inr,Welch t-test,Cohen's d,0.446
0,contract_type,chi-square,Cramer's V,0.410
2,internet_type,chi-square,Cramer's V,0.322
1,payment_method,chi-square,Cramer's V,0.303
3,senior_citizen,chi-square,Cramer's V,0.151


**Verdict.** Tenure (Cohen's d ≈ −0.85) and contract type (Cramér's V ≈ 0.41) are the load-bearing drivers; senior-citizen status is statistically real but small (V ≈ 0.15). Cramér's V and Cohen's d are on different scales, so the ranking is a heuristic, not an exact ordering.